src/model.py

```diff
@@ -214,71 +216,87 @@ def init_engine(self):
             self.vllm_engine_config = json.load(file)
 
         # Validate device and multi-processing settings are currently set based on model/configs.
-        self.validate_device_config()
+        self._validate_device_config()
 
         # Check for LoRA config and set it up if enabled
-        self.setup_lora()
-
-        # Create an AsyncLLMEngine from the config from JSON
-        aync_engine_args = AsyncEngineArgs(**self.vllm_engine_config)
-        self.llm_engine = AsyncLLMEngine.from_engine_args(aync_engine_args)
+        self._setup_lora()
 
-        # Create vLLM custom metrics
-        self.vllm_metrics = None
-        if (
-            self._get_bool_config_param("REPORT_CUSTOM_METRICS")
-            and not aync_engine_args.disable_log_stats
-        ):
-            try:
-                labels = {
-                    "model": self.args["model_name"],
-                    "version": self.args["model_version"],
-                }
-                # Add vLLM custom metrics
-                engine_config = self.llm_engine.engine.model_config
-                self.vllm_metrics = VllmStatLogger(
-                    labels, engine_config.max_model_len, self.logger
-                )
-                self.llm_engine.add_logger("triton", self.vllm_metrics)
-            except pb_utils.TritonModelException as e:
-                if "metrics not supported" in str(e):
-                    # Metrics are disabled at the server
-                    self.logger.log_info("[vllm] Metrics not supported")
-                else:
-                    raise e
+        # Create an AsyncEngineArgs from the config from JSON
+        self._aync_engine_args = AsyncEngineArgs(**self.vllm_engine_config)
 
-    def _get_bool_config_param(self, param_name: str) -> bool:
-        return (param_name in self.model_config["parameters"]) and (
-            self.model_config["parameters"][param_name]["string_value"].lower()
-            == "true"
+    def _init_engine(self):
+        # Run the engine in a separate thread running the AsyncIO event loop.
+        self._llm_engine = None
+        self._llm_engine_start_cv = threading.Condition()
+        self._llm_engine_shutdown_event = asyncio.Event()
+        self._event_thread = threading.Thread(
+            target=asyncio.run, args=(self._run_llm_engine(),)
```

Comment:
I found out why metrics were not shutting down properly, 
seems like `self._llm_engine_shutdown_event` being an `asyncio` event was accessed from different thread, which was causing issues. I managed to run into this with metrics OFF ZMQ route as well. 

Will include fix in a follow up pr

Reply:
> an asyncio event was accessed from different thread

This sounds a bit similar to the errors we're seeing on the L0_model_control_stress_*_vllm tests, in case your changes or investigation end up being helpful there

Reply:
seems to be unrelated 😭 Pipeline: 21686181, those still red

Reply:
those are not related, they are due to the Python AsyncIO client.

Reply:
Same debug approach may help on client side though